Latest version of BungeeCord gives an error

Plugin dont load 


18:27:34 [WARNING] Exception encountered when loading plugin: floodgate
java.lang.RuntimeException: Failed to inject the packet listener!
at org.geysermc.floodgate.FloodgatePlatform.enable(FloodgatePlatform.java:73)
at org.geysermc.floodgate.BungeePlugin.onEnable(BungeePlugin.java:64)
at net.md_5.bungee.api.plugin.PluginManager.enablePlugins(PluginManager.java:266)
at net.md_5.bungee.BungeeCord.start(BungeeCord.java:298)
at net.md_5.bungee.BungeeCordLauncher.main(BungeeCordLauncher.java:67)
at net.md_5.bungee.Bootstrap.main(Bootstrap.java:15)
Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.reflect.AccessibleObject.isAccessible()" because "accessibleObject" is null
at org.geysermc.floodgate.util.ReflectionUtils.makeAccessible(ReflectionUtils.java:570)
at org.geysermc.floodgate.util.ReflectionUtils.getValue(ReflectionUtils.java:249)
at org.geysermc.floodgate.util.ReflectionUtils.getCastedValue(ReflectionUtils.java:300)
at org.geysermc.floodgate.util.ReflectionUtils.castedStaticValue(ReflectionUtils.java:319)
at org.geysermc.floodgate.inject.bungee.BungeeInjector.inject(BungeeInjector.java:72)
at org.geysermc.floodgate.FloodgatePlatform.enable(FloodgatePlatform.java:71)
... 5 more

--- a/bungee/src/main/java/org/geysermc/floodgate/inject/bungee/BungeeInjector.java
+++ b/bungee/src/main/java/org/geysermc/floodgate/inject/bungee/BungeeInjector.java
@@ -28,15 +28,16 @@ package org.geysermc.floodgate.inject.bungee;
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelInitializer;
 import io.netty.channel.ChannelOutboundHandlerAdapter;
 import io.netty.channel.ChannelPromise;
 import java.lang.reflect.Field;
+import java.lang.reflect.Method;
 import lombok.Getter;
 import lombok.RequiredArgsConstructor;
 import net.md_5.bungee.netty.PipelineUtils;
 import net.md_5.bungee.protocol.MinecraftEncoder;
 import net.md_5.bungee.protocol.Varint21LengthFieldExtraBufPrepender;
-import net.md_5.bungee.protocol.Varint21LengthFieldPrepender;
 import org.geysermc.floodgate.api.logger.FloodgateLogger;
 import org.geysermc.floodgate.inject.CommonPlatformInjector;
 import org.geysermc.floodgate.util.BungeeReflectionUtils;
@@ -52,8 +53,10 @@ public final class BungeeInjector extends CommonPlatformInjector {
     @Override
     public void inject() {
         // Can everyone just switch to Velocity please :)
+        // :( ~BungeeCord Collaborator
 
         // Newer Bungee versions have a separate prepender for backend and client connections
+        // this field is not touched by client -> proxy
         Field serverFramePrepender =
                 ReflectionUtils.getField(PipelineUtils.class, "serverFramePrepender");
         if (serverFramePrepender != null) {
@@ -63,16 +66,25 @@ public final class BungeeInjector extends CommonPlatformInjector {
             BungeeReflectionUtils.setFieldValue(null, serverFramePrepender, customServerPrepender);
         }
 
-        Field framePrepender = ReflectionUtils.getField(PipelineUtils.class, "framePrepender");
-
-        // Required in order to inject into both Geyser <-> proxy AND proxy <-> server
-        // (Instead of just replacing the ChannelInitializer which is only called for
-        // player <-> proxy)
-        BungeeCustomPrepender customPrepender = new BungeeCustomPrepender(
-                this, ReflectionUtils.castedStaticValue(framePrepender)
+        // wrap the client -> proxy channel init because the framePrepender field was deleted
+        ChannelInitializer<Channel> original = PipelineUtils.SERVER_CHILD;
+        Field clientChannelInitField = ReflectionUtils.getField(
+                PipelineUtils.class, "SERVER_CHILD"
         );
-
-        BungeeReflectionUtils.setFieldValue(null, framePrepender, customPrepender);
+        Method initChannelMethod = ReflectionUtils.getMethod(
+                original.getClass(), "initChannel", Channel.class
+        );
+        ChannelInitializer<Channel> wrapper = new ChannelInitializer<Channel>() {
+            @Override
+            protected void initChannel(Channel channel) {
+                ReflectionUtils.invoke(original, initChannelMethod, channel);
+                channel.pipeline().addBefore(
+                        PipelineUtils.FRAME_DECODER, BUNGEE_INIT,
+                        new BungeeClientToProxyInjectInitializer(BungeeInjector.this)
+                );
+            }
+        };
+        BungeeReflectionUtils.setFieldValue(null, clientChannelInitField, wrapper);
 
         injected = true;
     }
@@ -103,31 +115,6 @@ public final class BungeeInjector extends CommonPlatformInjector {
 
         injectAddonsCall(channel, !clientToProxy);
         addInjectedClient(channel);
-    }
-
-    @RequiredArgsConstructor
-    private static final class BungeeCustomPrepender extends Varint21LengthFieldPrepender {
-        private final BungeeInjector injector;
-        private final Varint21LengthFieldPrepender original;
-
-        @Override
-        public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
-            original.handlerAdded(ctx);
-            // The Minecraft encoder being in the pipeline isn't present until later
-
-            if (ctx.channel().parent() != null) {
-                // Client <-> Proxy
-                ctx.pipeline().addBefore(
-                        PipelineUtils.FRAME_DECODER, BUNGEE_INIT,
-                        new BungeeClientToProxyInjectInitializer(injector)
-                );
-            } else {
-                // Proxy <-> Server
-                ctx.pipeline().addLast(
-                        BUNGEE_INIT, new BungeeProxyToServerInjectInitializer(injector)
-                );
-            }
-        }
     }
 
     @RequiredArgsConstructor

--- a/bungee/src/main/java/org/geysermc/floodgate/inject/bungee/BungeeInjector.java
+++ b/bungee/src/main/java/org/geysermc/floodgate/inject/bungee/BungeeInjector.java
@@ -38,6 +38,7 @@ import lombok.RequiredArgsConstructor;
 import net.md_5.bungee.netty.PipelineUtils;
 import net.md_5.bungee.protocol.MinecraftEncoder;
 import net.md_5.bungee.protocol.Varint21LengthFieldExtraBufPrepender;
+import net.md_5.bungee.protocol.Varint21LengthFieldPrepender;
 import org.geysermc.floodgate.api.logger.FloodgateLogger;
 import org.geysermc.floodgate.inject.CommonPlatformInjector;
 import org.geysermc.floodgate.util.BungeeReflectionUtils;
@@ -66,25 +67,39 @@ public final class BungeeInjector extends CommonPlatformInjector {
             BungeeReflectionUtils.setFieldValue(null, serverFramePrepender, customServerPrepender);
         }
 
-        // wrap the client -> proxy channel init because the framePrepender field was deleted
-        ChannelInitializer<Channel> original = PipelineUtils.SERVER_CHILD;
-        Field clientChannelInitField = ReflectionUtils.getField(
-                PipelineUtils.class, "SERVER_CHILD"
-        );
-        Method initChannelMethod = ReflectionUtils.getMethod(
-                original.getClass(), "initChannel", Channel.class
-        );
-        ChannelInitializer<Channel> wrapper = new ChannelInitializer<Channel>() {
-            @Override
-            protected void initChannel(Channel channel) {
-                ReflectionUtils.invoke(original, initChannelMethod, channel);
-                channel.pipeline().addBefore(
-                        PipelineUtils.FRAME_DECODER, BUNGEE_INIT,
-                        new BungeeClientToProxyInjectInitializer(BungeeInjector.this)
-                );
-            }
-        };
-        BungeeReflectionUtils.setFieldValue(null, clientChannelInitField, wrapper);
+        // for backwards compatibility
+        Field framePrepender = ReflectionUtils.getField(PipelineUtils.class, "framePrepender");
+        if (framePrepender != null) {
+            logger.warn("You are running an old version of BungeeCord consider updating to a newer version");
+            // Required in order to inject into both Geyser <-> proxy AND proxy <-> server
+            // (Instead of just replacing the ChannelInitializer which is only called for
+            // player <-> proxy)
+            BungeeCustomPrepender customPrepender = new BungeeCustomPrepender(
+                    this, ReflectionUtils.castedStaticValue(framePrepender)
+            );
+
+            BungeeReflectionUtils.setFieldValue(null, framePrepender, customPrepender);
+        } else {
+            // wrap the client -> proxy channel init because the framePrepender field was deleted
+            ChannelInitializer<Channel> original = PipelineUtils.SERVER_CHILD;
+            Field clientChannelInitField = ReflectionUtils.getField(
+                    PipelineUtils.class, "SERVER_CHILD"
+            );
+            Method initChannelMethod = ReflectionUtils.getMethod(
+                    original.getClass(), "initChannel", Channel.class
+            );
+            ChannelInitializer<Channel> wrapper = new ChannelInitializer<Channel>() {
+                @Override
+                protected void initChannel(Channel channel) {
+                    ReflectionUtils.invoke(original, initChannelMethod, channel);
+                    channel.pipeline().addBefore(
+                            PipelineUtils.FRAME_DECODER, BUNGEE_INIT,
+                            new BungeeClientToProxyInjectInitializer(BungeeInjector.this)
+                    );
+                }
+            };
+            BungeeReflectionUtils.setFieldValue(null, clientChannelInitField, wrapper);
+        }
 
         injected = true;
     }
@@ -115,6 +130,31 @@ public final class BungeeInjector extends CommonPlatformInjector {
 
         injectAddonsCall(channel, !clientToProxy);
         addInjectedClient(channel);
+    }
+
+    @RequiredArgsConstructor
+    private static final class BungeeCustomPrepender extends Varint21LengthFieldPrepender {
+        private final BungeeInjector injector;
+        private final Varint21LengthFieldPrepender original;
+
+        @Override
+        public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+            original.handlerAdded(ctx);
+            // The Minecraft encoder being in the pipeline isn't present until later
+
+            if (ctx.channel().parent() != null) {
+                // Client <-> Proxy
+                ctx.pipeline().addBefore(
+                        PipelineUtils.FRAME_DECODER, BUNGEE_INIT,
+                        new BungeeClientToProxyInjectInitializer(injector)
+                );
+            } else {
+                // Proxy <-> Server
+                ctx.pipeline().addLast(
+                        BUNGEE_INIT, new BungeeProxyToServerInjectInitializer(injector)
+                );
+            }
+        }
     }
 
     @RequiredArgsConstructor

